File: code/environment_models/src/Airy.cpp
/*
 * Airy.cpp
 *
 *  Created on: Aug 1, 2014
 *      Author: cady
 */

#include "Airy.hpp"
#include "InternalErrorException.hpp"
#include "InvalidInputException.hpp"
#include "discretize.hpp"

#include <vector>
#include <ssc/macros.hpp>



Airy::Airy(const DiscreteDirectionalWaveSpectrum& spectrum_, const double constant_random_phase) : WaveModel(spectrum_, constant_random_phase)
{
}

Airy::Airy(const DiscreteDirectionalWaveSpectrum& spectrum_, const int random_number_generator_seed) : WaveModel(spectrum_, random_number_generator_seed)
{
}


double Airy::evaluate_rao(
        const double x,                           //!< x-position of the RAO's calculation point in the NED frame (in meters)
        const double y,                           //!< y-position of the RAO's calculation point in the NED frame (in meters)
        const double t,                           //!< Current time instant (in seconds)
        const std::vector<double>& rao_module_for_each_frequency_and_incidence,    //!< "Flattened" matrix containing the module of the RAO for each angular frequency omega and each wave incidence beta
        const std::vector<double>& rao_phase_for_each_frequency_and_incidence      //!< "Flattened" matrix containing the phase of the RAO for each angular frequency omega and each wave incidence beta
        ) const
{
    double F = 0;
    const size_t nb_of_omegas_x_nb_of_directions = rao_module_for_each_frequency_and_incidence.size();
    if (nb_of_omegas_x_nb_of_directions != flat_spectrum.k.size())
    {
        THROW(__PRETTY_FUNCTION__, InternalErrorException, "Number of angular frequencies times number of incidences in HDB RAO is " << nb_of_omegas_x_nb_of_directions << ", which does not match spectrum size (" << flat_spectrum.k.size() << " (omega,psi) pairs)");
    }
    for (size_t i = 0 ; i < nb_of_omegas_x_nb_of_directions ; ++i) // For each (omega,beta) pair
    {
        const double rao_amplitude = rao_module_for_each_frequency_and_incidence[i] * flat_spectrum.a[i];
        const double omega_t = flat_spectrum.omega[i] * t;
        const double k_xCosPsi_ySinPsi = flat_spectrum.k[i] * (x * flat_spectrum.cos_psi[i] + y * flat_spectrum.sin_psi[i]);
        const double theta = flat_spectrum.phase[i];
        F -= rao_amplitude * sin(-omega_t + k_xCosPsi_ySinPsi + theta + rao_phase_for_each_frequency_and_incidence[i]);
    }
    return F;
}

std::vector<double> Airy::elevation(
    const std::vector<double> &x, //!< x-positions in the NED frame (in meters)
    const std::vector<double> &y, //!< y-positions in the NED frame (in meters)
    const double t                //!< Current time instant (in seconds)
    ) const
{
    std::vector<double> zeta(x.size());
    const size_t n = flat_spectrum.psi.size();

    for (size_t j = 0; j < zeta.size(); ++j) {
        for (size_t i = 0 ; i < n ; ++i)
        {
            const double a = flat_spectrum.a[i];
            const double omega_t = flat_spectrum.omega[i] * t;
            const double k_xCosPsi_ySinPsi = flat_spectrum.k[i] * (x.at(j) * flat_spectrum.cos_psi[i] + y.at(j) * flat_spectrum.sin_psi[i]);
            const double theta = flat_spectrum.phase[i];
            zeta.at(j) -= a * sin(-omega_t + k_xCosPsi_ySinPsi + theta);
        }
    }

    return zeta;
}

std::vector<double> Airy::dynamic_pressure(
    const double rho,               //!< water density (in kg/m^3)
    const double g,                 //!< gravity (in m/s^2)
    const std::vector<double> &x,   //!< x-positions in the NED frame (in meters)
    const std::vector<double> &y,   //!< y-positions in the NED frame (in meters)
    const std::vector<double> &z,   //!< z-positions in the NED frame (in meters)
    const std::vector<double> &eta, //!< Wave elevations at (x,y) in the NED frame (in meters)
    const double t                  //!< Current time instant (in seconds)
    ) const
{
    std::vector<double> p(x.size(), 0);

    for (size_t j = 0; j < p.size(); ++j)
    {
        if (std::isnan(z[j]))
        {
            THROW(__PRETTY_FUNCTION__, InternalErrorException, "z (value to rescale, in meters) was NaN");
        }
        if (std::isnan(eta[j]))
        {
            THROW(__PRETTY_FUNCTION__, InternalErrorException, "eta (wave height, in meters) was NaN");
        }

        if (z[j] < eta[j])
        {
            p[j] = 0;
        }
        else
        {
            const size_t n = flat_spectrum.psi.size();
            for (size_t i = 0; i < n; ++i)
            {
                const double a = flat_spectrum.a[i];
                const double k = flat_spectrum.k[i];
                const double omega_t = flat_spectrum.omega[i] * t;
                const double pdyn_fact = flat_spectrum.pdyn_factor(k, z[j], eta[j]);
                const double k_xCosPsi_ySinPsi = k * (x[j] * flat_spectrum.cos_psi[i] + y[j] * flat_spectrum.sin_psi[i]);
                const double theta = flat_spectrum.phase[i];
                p[j] += a * pdyn_fact * sin(-omega_t + k_xCosPsi_ySinPsi + theta);
            }
            p[j] *= rho * g;
        }
        
    }
    return p;
}

ssc::kinematics::PointMatrix Airy::orbital_velocity(
        const double g,                //!< gravity (in m/s^2)
        const std::vector<double>& x,  //!< x-positions in the NED frame (in meters)
        const std::vector<double>& y,  //!< y-positions in the NED frame (in meters)
        const std::vector<double>& z,  //!< z-positions in the NED frame (in meters)
        const double t,                //!< Current time instant (in seconds)
        const std::vector<double>& eta //!< Wave heights at x,y,t (in meters)
        ) const
{
    ssc::kinematics::PointMatrix M("NED", x.size());
    for (size_t point_index = 0; point_index < x.size(); ++point_index) {

        if (z.at(point_index) < eta.at(point_index)) 
        {
            M.m(0, point_index) = 0;
            M.m(1, point_index) = 0;
            M.m(2, point_index) = 0;
        } else {
            const size_t n = flat_spectrum.psi.size();
            double u = 0;
            double v = 0;
            double w = 0;
            for (size_t i = 0 ; i < n ; ++i)
            {
                const double omega = flat_spectrum.omega[i];
                const double k = flat_spectrum.k[i];
                const double pdyn_factor = flat_spectrum.pdyn_factor(k,z[point_index],0); // No stretching for the orbital velocity
                const double pdyn_factor_sh = flat_spectrum.pdyn_factor_sh(k,z[point_index],0); // No stretching for the orbital velocity
                const double k_xCosPsi_ySinPsi = k * (x[point_index] * flat_spectrum.cos_psi[i] + y[point_index] * flat_spectrum.sin_psi[i]);
                const double theta = -omega * t + k_xCosPsi_ySinPsi + flat_spectrum.phase[i];
                const double cos_theta = cos(theta);
                const double sin_theta = sin(theta);
                const double a_k_omega = flat_spectrum.a[i] * k / omega;
                const double a_k_omega_pdyn_factor_sin_theta = a_k_omega * pdyn_factor * sin_theta;
                u += a_k_omega_pdyn_factor_sin_theta * flat_spectrum.cos_psi[i];
                v += a_k_omega_pdyn_factor_sin_theta * flat_spectrum.sin_psi[i];
                w += a_k_omega * pdyn_factor_sh * cos_theta;
            }
            M.m(0, point_index) = u * g;
            M.m(1, point_index) = v * g;
            M.m(2, point_index) = w * g;
        }
    }
    return M;

}

File: grpc_tests/waves/airy.py
"""Airy wave model. As implemented in xdyn."""

import logging
import math
import yaml
import numpy as np
import waves


SERVICE_NAME = "airy"

logging.basicConfig(
    format='%(asctime)s,%(msecs)d ['
    + SERVICE_NAME
    + '] - %(levelname)-4s [%(filename)s:%(lineno)d] %(message)s',
    datefmt='%d-%m-%Y:%H:%M:%S')
LOGGER = logging.getLogger(__name__)
LOGGER.setLevel(logging.INFO)


def pdyn_factor(k, z, eta):
    """exp(-kz) factor used for the dynamic pressure calculations.

    Parameters
    ----------
    k : float
        Wave number (in metres^-1).
    z : float
        Position (in meters) at which we want the elevation. Projected on
        the Z-axis of the Earth-centered, Earth-fixed North-East-Down
        reference frame.
    eta : float
        Wave elevation (in meters) in the North-East-Down reference frame.

    Returns
    -------
    float
        dynamic pressure factor.

    """
    return 0 if (eta != 0 and z < eta) else math.exp(-k * z)


class Airy(waves.AbstractWaveModel):
    """Linear irregular waves in infinite depth.

    Using JONSWAP spectrum and single propagation direction. No stretching.
    """

    def __init__(self):
        """Constructor."""
        self.psi0 = None
        self.jonswap_parameters = {'sigma_a': 0.07, 'sigma_b': 0.09}
        self.directional_spectrum = {}

    def set_parameters(self, parameters):
        """Initialize the wave model with YAML parameters.

        Parameters
        ----------
        parameters : string
            YAML string containing the parameters of this model. The YAML
            should have the following form:
                waves propagating to: 90
                Hs: 5
                Tp: 15
                gamma: 1.2
                omega: [1,2,3]
            `waves propagating to`: direction, in degrees,  the waves are
                                    propagating to. 0 for waves coming from the
                                    South and propagating to the North, 90 deg
                                    for waves coming from the East and
                                    propagating to the West.
            `Hs`: significant wave height in meters, used in the JONSWAP
                  spectrum.
            `Tp`: peak wave period, in seconds, used in the JONSWAP spectrum.
            `gamma`: JONSWAP shape parameter
            `omegas`: discretization of the angular frequency. In rad/s.
            `psis`: discretization of the incidence. In rad.


        Returns
        -------
        Nothing

        """
        param = yaml.safe_load(parameters)
        self.jonswap_parameters['t_p'] = param['Tp']
        self.jonswap_parameters['gamma'] = param['gamma']
        self.directional_spectrum['omega'] = param['omega']
        self.directional_spectrum['psi'] = \
            [param['waves propagating to'] * math.pi / 180.0]
        self.jonswap_parameters['hs_square'] = param['Hs'] ** 2
        self.jonswap_parameters['omega0'] = 2 * math.pi / param['Tp']
        self.jonswap_parameters['coeff'] = 1 - 0.287 * math.log(param['gamma'])
        self.directional_spectrum['si'] = [self.jonswap(omega) for omega in
                                           param['omega']]
        self.directional_spectrum['dj'] = [1]
        self.directional_spectrum['k'] = [omega * omega / 9.81 for omega in
                                          param['omega']]
        phases = np.random.uniform(low=0,
                                   high=2 * math.pi,
                                   size=(len(param['omega']),))
        self.directional_spectrum['phase'] = phases

    def jonswap(self, omega):
        r"""Joint North Sea Project spectrum.

        ```math
        S(\omega)=(1-0.287 \log(\gamma))\frac{5}{16}\frac{\alpha}{\omega}H_S^2
        e^{-1.25\left(\frac{\omega_0}{\omega}\right)^4}\gamma^r
        ```

        with

        ```math
        r=e^{-0.5\left(\frac{\omega-\omega_0}{\sigma\omega_0}\right)^2}
        ```

        and

        ```math
        \sigma=\left\{\begin{array}{l}0.07,\omega\leq\omega_0\\0.09,
        \omega>\omega_0\end{array}\right.
        ```
        """
        sigma_a = self.jonswap_parameters['sigma_a']
        sigma_b = self.jonswap_parameters['sigma_b']
        omega0 = self.jonswap_parameters['omega0']
        hs_square = self.jonswap_parameters['hs_square']
        coeff = self.jonswap_parameters['coeff']
        gamma = self.jonswap_parameters['gamma']
        sigma = sigma_a if omega <= omega0 else sigma_b
        ratio = omega0 / omega
        alpha = ratio ** 4
        awm_5 = coeff * 5.0 / 16.0 * alpha / omega * hs_square
        bwm_4 = 1.25 * alpha
        kappa = (omega - omega0) / (sigma * omega0)
        return awm_5 * math.exp(-bwm_4) * math.pow(gamma,
                                                   math.exp(-0.5*kappa ** 2))

    def elevation(self, x, y, t):
        """Calculate the elevations of the free surface at any point in time.

        Parameters
        ----------
        x : float
            Position (in meters) at which we want the elevation. Projected on
            the X-axis of the Earth-centered, Earth-fixed North-East-Down
            reference frame.
        y : float
            Position (in meters) at which we want the elevation. Projected on
            the Y-axis of the Earth-centered, Earth-fixed North-East-Down
            reference frame.
        t : float
            Simulation time (in seconds). All values are calculated at that
            instant. The documentation of each waves model should describe how
            the phases are defined.

        Returns
        -------
        float
            Free surface height along the Z-axis (oriented downwards) in
            meters.

        """
        zeta = 0
        dir_spec = self.directional_spectrum
        psi = dir_spec['psi'][0]
        for s_i, k, omega, phase in zip(dir_spec['si'],
                                        dir_spec['k'],
                                        dir_spec['omega'],
                                        dir_spec['phase']):
            k_x_cos_psi_y_sin_psi = k * (x * math.cos(psi) + y * math.sin(psi))
            zeta -= s_i * math.sin(-omega*t + k_x_cos_psi_y_sin_psi + phase)
        return zeta

    def dynamic_pressure(self, x, y, z, t):
        """Calculate the dynamic pressure due to waves.

        Parameters
        ----------
        x : float
            Position (in meters) at which we want the elevation. Projected on
            the X-axis of the Earth-centered, Earth-fixed North-East-Down
            reference frame.
        y : float
            Position (in meters) at which we want the elevation. Projected on
            the Y-axis of the Earth-centered, Earth-fixed North-East-Down
            reference frame.
        z : float
            Position (in meters) at which we want the elevation. Projected on
            the Z-axis of the Earth-centered, Earth-fixed North-East-Down
            reference frame.
        t : float
            Simulation time (in seconds). All values are calculated at that
            instant. The documentation of each waves model should describe how
            the phases are defined.

        Returns
        -------
        float
            Dynamic pressure (in Pascal) at (x,y,z,t).

        """
        dir_spec = self.directional_spectrum
        eta = self.elevation(x, y, t)
        acc = 0
        psi = dir_spec['psi'][0]
        cos_psi = math.cos(psi)
        sin_psi = math.sin(psi)
        cos_psi_y_sin_psi = (x * cos_psi + y * sin_psi)
        for s_i, k, omega, phase in zip(dir_spec['si'],
                                        dir_spec['k'],
                                        dir_spec['omega'],
                                        dir_spec['phase']):
            acc -= s_i * pdyn_factor(k, z, eta) * math.sin(-omega * t
                                                           + k * cos_psi_y_sin_psi
                                                           + phase)
        return 1000 * 9.81 * acc

    def orbital_velocity(self, x, y, z, t):
        """Calculate the orbital velocity of the wave particles.

        Parameters
        ----------
        x : float
            Position (in meters) at which we want the elevation. Projected on
            the X-axis of the Earth-centered, Earth-fixed North-East-Down
            reference frame.
        y : float
            Position (in meters) at which we want the elevation. Projected on
            the Y-axis of the Earth-centered, Earth-fixed North-East-Down
            reference frame.
        z : float
            Position (in meters) at which we want the elevation. Projected on
            the Z-axis of the Earth-centered, Earth-fixed North-East-Down
            reference frame.
        t : float
            Simulation time (in seconds). All values are calculated at that
            instant. The documentation of each waves model should describe how
            the phases are defined.

        Returns
        -------
        dict
            Should contain the following fields:
            - vx (float): Projection on the X-axis of the Earth-centered,
                          Earth-fixed North-East-Down reference frame of the
                          velocity of each wave partical relative to the
                          ground. In meters per second.
            - vy (float): Projection on the Y-axis of the Earth-centered,
                          Earth-fixed North-East-Down reference frame of the
                          velocity of each wave partical relative to the
                          ground. In meters per second.
            - vz (float): Projection on the Z-axis of the Earth-centered,
                          Earth-fixed North-East-Down reference frame of the
                          velocity of each wave partical relative to the
                          ground. In meters per second.

        """
        dir_spec = self.directional_spectrum
        eta = self.elevation(x, y, t)
        v_x = 0
        v_y = 0
        v_z = 0
        psi = dir_spec['psi'][0]
        x_cos_psi_y_sin_psi = (x * math.cos(psi) + y * math.sin(psi))
        for s_i, k, omega, phase in zip(dir_spec['si'],
                                        dir_spec['k'],
                                        dir_spec['omega'],
                                        dir_spec['phase']):
            pdyn_factor = self.pdyn_factor(k, z, eta)
            pdyn_factor_sh = pdyn_factor
            theta = -omega * t + k * x_cos_psi_y_sin_psi + phase
            cos_theta = math.cos(theta)
            sin_theta = math.sin(theta)
            a_k_omega = s_i * k / omega
            a_k_omega_pdyn_factor_sin_theta = a_k_omega * pdyn_factor \
                * sin_theta
            v_x += a_k_omega_pdyn_factor_sin_theta * math.cos(psi)
            v_y += a_k_omega_pdyn_factor_sin_theta * math.sin(psi)
            v_z += a_k_omega * pdyn_factor_sh * cos_theta

        return {'vx': v_x, 'vy': v_y, 'vz': v_z}

    def angular_frequencies_for_rao(self):
        """Get angular frequencies the wave spectrum is discretized at.

        Returns
        -------
        list of floats
            Angular frequencies the spectrum was discretized at (in rad/s).
            Used, for example, when interpolating the wave RAOs.

        """
        return self.directional_spectrum['omegas']

    def directions_for_rao(self):
        """Get the incidences the directional spreading is discretized at.

        Returns
        -------
        list of floats
          Wave incidences the spectrum was discretized at (in rad).
          0° is for waves coming from the North.
          90° for waves coming from the East.
          Used, for example, when interpolating the wave RAOs.

        """
        return self.directional_spectrum['psis']

    def spectrum(self, x, y, t):
        """Linear spectrum that can be used by xdyn's diffraction module.

        Parameters
        ----------
        x : float
            Position (in meters) at which we want the (linearized) spectrum.
            Projected on the X-axis of the Earth-centered, Earth-fixed
            North-East-Down reference frame.
        y : float
            Position (in meters) at which we want the (linearized) spectrum.
            Projected on the Y-axis of the Earth-centered, Earth-fixed
            North-East-Down reference frame.
        t : float
            Simulation time (in seconds). Date at which we want the
            (linearized) spectrum.

        Returns
        -------
        dict
            Should contain the following fields:
            - si (list of floats): Discretized spectral density for each
              omega (should therefore be the same size as omega).
              In s m^2/rad.
            - dj (list of floats): Spatial spreading for each psi (should
              therefore be the same size as psi. In 1/rad.
            - omega (list of floats): Angular frequencies the spectrum was
              discretized at. In rad/s.
            - psi (list of floats): Directions between 0 & 2pi the spatial
              spreading was discretized at. In rad.
            - k (list of floats): Discretized wave number for each
              frequency (should therefore be the same size as omega).
              In rad/m.
            - phase (list of lists): Random phases, for each
              (direction,frequency) couple (but time invariant), should
              have the same size as psi.
              Each element should have the same size as omega. In radian.

        """
        return self.directional_spectrum


if __name__ == '__main__':
    LOGGER.info('Starting Airy waves server...')
    waves.serve(Airy())
    LOGGER.info('Stopped Airy waves server')
